serializers/avro/src/main/java/io/pravega/schemaregistry/serializer/avro/impl/AvroDeserializer.java

```diff
@@ -19,48 +19,55 @@
 import io.pravega.schemaregistry.serializer.shared.impl.SerializerConfig;
 import org.apache.avro.Schema;
 import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.DatumReader;
 import org.apache.avro.io.DecoderFactory;
 import org.apache.avro.reflect.ReflectDatumReader;
 import org.apache.avro.specific.SpecificDatumReader;
 import org.apache.avro.specific.SpecificRecordBase;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.nio.ByteBuffer;
 import java.util.concurrent.ConcurrentHashMap;
 
 class AvroDeserializer<T> extends AbstractDeserializer<T> {
-    private final AvroSchema<T> avroSchema;
-    private final ConcurrentHashMap<SchemaInfo, Schema> knownSchemas;
+    private final ConcurrentHashMap<ByteBuffer, DatumReader<T>> knownSchemas;
+    private final boolean specific;
+    private final Schema readerSchema;
 
     AvroDeserializer(String groupId, SchemaRegistryClient client,
                      AvroSchema<T> schema,
                      SerializerConfig.Decoders decoder, EncodingCache encodingCache) {
         super(groupId, client, schema, false, decoder, encodingCache, true);
         Preconditions.checkNotNull(schema);
-        this.avroSchema = schema;
         this.knownSchemas = new ConcurrentHashMap<>();
+        specific = SpecificRecordBase.class.isAssignableFrom(schema.getTClass());
+        readerSchema = schema.getSchema();
+        ByteBuffer schemaData = schema.getSchemaInfo().getSchemaData();
+        createDatumReader(schemaData, readerSchema);
     }
 
     @Override
     public final T deserialize(InputStream inputStream, SchemaInfo writerSchemaInfo, SchemaInfo readerSchemaInfo) throws IOException {
         Preconditions.checkNotNull(writerSchemaInfo);
-        Schema writerSchema;
-        if (knownSchemas.containsKey(writerSchemaInfo)) {
-            writerSchema = knownSchemas.get(writerSchemaInfo);
-        } else {
+        DatumReader<T> datumReader  = knownSchemas.get(writerSchemaInfo.getSchemaData());
+        if (datumReader == null) {
```

Comment:
we can make one improvement here -- we can call `knownSchemas.compute` so in case this method is called concurrently, we avoid parallel computations.

Reply:
Done